src/psyclone/psyGen.py

```diff
@@ -2988,6 +3025,198 @@ def ast(self):
+        return original + tag + suffix
+
+    def update(self):
+        '''
+        Writes the (transformed) AST of this kernel to file and resets the
+        'modified' flag to False. By default, the kernel is re-named
+        so as to be unique within the kernel output directory stored
+        within the configuration object. If config.kernel_naming is
+        "single" then no re-naming and output is performed if there is
+        already a transformed copy of the kernel in the output dir.
+
+        :raises GenerationError: if config.kernel_naming == "single" and a \
+                                 different, transformed version of this \
+                                 kernel is already in the output directory.
+        '''
+        import os
+        from psyclone.line_length import FortLineLength
+
+        # If this kernel has not been transformed we do nothing
+        if not self.modified:
+            return
+
+        # Remove any "_mod" if the file follows the PSyclone naming convention
+        orig_mod_name = self.module_name[:]
+        if orig_mod_name.endswith("_mod"):
+            old_base_name = orig_mod_name[:-4]
+        else:
+            old_base_name = orig_mod_name[:]
+
+        # We could create a hash of a string built from the name of the
+        # Algorithm (module), the name/position of the Invoke and the
+        # index of this kernel within that Invoke. However, that creates
+        # a very long name so we simply ensure that kernel names are unique
+        # within the user-supplied kernel-output directory.
+        name_idx = -1
+        fdesc = None
+        while not fdesc:
+            name_idx += 1
+            new_suffix = "_{0}".format(name_idx)
+            new_name = old_base_name + new_suffix + "_mod.f90"
+            try:
+                # Atomically attempt to open the new kernel file (in case
+                # this is part of a parallel build)
+                fdesc = os.open(
+                    os.path.join(Config.get().kernel_output_dir, new_name),
+                    os.O_CREAT | os.O_WRONLY | os.O_EXCL)
+            except (OSError, IOError):
+                # The os.O_CREATE and os.O_EXCL flags in combination mean
+                # that open() raises an error if the file exists
+                if Config.get().kernel_naming == "single":
+                    # If the kernel-renaming scheme is such that we only ever
+                    # create one copy of a transformed kernel then we're done
+                    break
+                continue
+
+        # Use the suffix we have determined to rename all relevant quantities
+        # within the AST of the kernel code
+        self._rename_ast(new_suffix)
+
+        # Generate the Fortran for this transformed kernel, ensuring that
+        # we limit the line lengths
+        fll = FortLineLength()
+        new_kern_code = fll.process(str(self.ast))
+
+        if not fdesc:
```

Comment:
Does it make sense (and is it possible) to check that with "unique" as the option we don't output a new file name if the content is the same?

Reply:
If you mean from within the test suite then I'm not sure it's possible without diving in to checking file-creation times. We test that an error is raised if the kernel does already exist and is different. Testing that we don't create something identical to something that we already have is difficult and won't alter correctness (since either way, we end up with the thing that we want).

Reply:
OK.